src/wallet/gtest/test_wallet.cpp

```diff
@@ -806,6 +806,127 @@ TEST(WalletTests, GetConflictedSaplingNotes) {
+
+    EXPECT_EQ(-1, chainActive.Height());
+    CBlock block;
+    block.vtx.push_back(wtx);
+    block.hashMerkleRoot = block.BuildMerkleTree();
+    auto blockHash = block.GetHash();
+    CBlockIndex fakeIndex {block};
+    fakeIndex.hashFinalOrchardRoot = orchardTree.root();
+    mapBlockIndex.insert(std::make_pair(blockHash, &fakeIndex));
+    chainActive.SetTip(&fakeIndex);
+    EXPECT_TRUE(chainActive.Contains(&fakeIndex));
+    EXPECT_EQ(0, chainActive.Height());
+
+    // Simulate SyncTransaction which calls AddToWalletIfInvolvingMe
+    auto orchardTxMeta = wallet.GetOrchardWallet().AddNotesIfInvolvingMe(wtx);
+    ASSERT_TRUE(orchardTxMeta.has_value());
+    EXPECT_FALSE(orchardTxMeta.value().empty());
+    wtx.SetOrchardTxMeta(orchardTxMeta.value());
+    wtx.SetMerkleBranch(block);
+    wallet.LoadWalletTx(wtx);
+
+    // Simulate receiving new block and ChainTip signal
+    wallet.IncrementNoteWitnesses(Params().GetConsensus(),&fakeIndex, &block, sproutTree, saplingTree, true);
+    wallet.UpdateSaplingNullifierNoteMapForBlock(&block);
+
+    // Fetch the Orchard note so we can spend it.
+    std::vector<SproutNoteEntry> sproutEntries;
+    std::vector<SaplingNoteEntry> saplingEntries;
+    std::vector<OrchardNoteMetadata> orchardEntries;
+    wallet.GetFilteredNotes(sproutEntries, saplingEntries, orchardEntries, std::nullopt, -1);
+    EXPECT_EQ(0, sproutEntries.size());
+    EXPECT_EQ(0, saplingEntries.size());
+    EXPECT_EQ(1, orchardEntries.size());
+
+    // Generate another recipient
+    libzcash::diversifier_index_t j2(1);
+    auto recipient2 = ivk.Address(j2);
+
+    // Generate tx to spend note A
+    auto noteToSpend = std::move(wallet.GetOrchardSpendInfo(orchardEntries)[0]);
+    auto builder2 = TransactionBuilder(consensusParams, 2, orchardTree.root());
+    builder2.AddOrchardSpend(std::move(noteToSpend.first), std::move(noteToSpend.second));
+    auto tx2 = builder2.Build().GetTxOrThrow();
+    CWalletTx wtx2 {&wallet, tx2};
+
+    // Generate conflicting tx to spend note A
+    auto noteToSpend2 = std::move(wallet.GetOrchardSpendInfo(orchardEntries)[0]);
+    auto builder3 = TransactionBuilder(consensusParams, 2, orchardTree.root());
+    builder3.AddOrchardSpend(std::move(noteToSpend2.first), std::move(noteToSpend2.second));
+    auto tx3 = builder3.Build().GetTxOrThrow();
+    CWalletTx wtx3 {&wallet, tx3};
+
+    auto hash = wtx.GetHash();
+    auto hash2 = wtx2.GetHash();
+    auto hash3 = wtx3.GetHash();
+
+    // No conflicts for no spends (wtx is currently the only transaction in the wallet)
+    EXPECT_EQ(0, wallet.GetConflicts(hash).size());
+
+    // No conflicts for one spend
+    auto orchardTxMeta2 = wallet.GetOrchardWallet().AddNotesIfInvolvingMe(wtx2);
+    ASSERT_TRUE(orchardTxMeta2.has_value());
+    EXPECT_FALSE(orchardTxMeta2.value().empty());
+    wtx2.SetOrchardTxMeta(orchardTxMeta2.value());
+    wallet.LoadWalletTx(wtx2);
+    EXPECT_EQ(0, wallet.GetConflicts(hash).size());
```

Comment:
This is checking the wrong thing. `wtx2` spends the outputs of `wtx`; it does not spend the inputs of `wtx`. This should do what we do in the Sapling test:
```suggestion
    EXPECT_EQ(0, wallet.GetConflicts(hash2).size());
```

Reply:
`GetConflicts(tx)` uses `orchard_wallet_get_potential_spends` for each action in the `tx`'s Orchard bundle.

`orchard_wallet_get_potential_spends` checks the spentness of the output of the action:
```rust
    if let Some(inpoints) = wallet
        .wallet_received_notes
        .get(&txid)
        .and_then(|txnotes| txnotes.decrypted_notes.get(&action_idx.try_into().unwrap()))
        .and_then(|dnote| wallet.key_store.get_nullifier(&dnote.note))
        .and_then(|nf| wallet.potential_spends.get(&nf))
    {
        for inpoint in inpoints {
            unsafe { (push_cb.unwrap())(result, inpoint.txid.as_ref()) };
        }
    }
```
We could modify this to check the spentness of the input instead, in which case the test would work as you expected:

```rust
    if let Some(inpoints) = wallet
        .wallet_received_notes
        .get(&txid)
        .and_then(|txnotes| txnotes.decrypted_notes.get(&action_idx.try_into().unwrap()))
        .and_then(|dnote| dnote.note.rho)
        .and_then(|nf| wallet.potential_spends.get(&nf))
    {
        for inpoint in inpoints {
            unsafe { (push_cb.unwrap())(result, inpoint.txid.as_ref()) };
        }
    }
```

However, that would change the semantics of `IsOrchardSpent`, which is the other method that uses `orchard_wallet_get_potential_spends`.

Reply:
The bug here is the use of `orchardWallet.GetPotentialSpends(op)` in `CWallet::GetConflicts`; the multiple layers of indirection, combined with the fact that `orchardWallet.GetPotentialSpends(op)` is not documented, meant that I missed this inconsistency during my review of #5649. Additionally, `CWallet::IsOrchardSpent` takes an outpoint, compared to `CWallet::IsSproutSpent` and `CWallet::IsSaplingSpent` which take a nullifier directly. The two are "equivalent", but an outpoint has an extra layer of indirection due to the zero knowledge layer.

What we need is a separate method on `OrchardWallet`, that takes a nullifier and returns the txids of transactions that attempt to spend it. That removes the layer of indirection that causes this bug. Note that for transparent, you _can_ use an outpoint in this way, because there is no zero knowledge layer: the outpoint _is_ what appears in the spend side of the transaction.